### ISSUE TITLE
Cognito user `sub` attribute should appear in `cognito-idp list-users` output

### ISSUE THREAD

**Issue by soulshake** [localstack/localstack#2963]
# Type of request: This is a ...

[x] bug report
[ ] feature request

# Detailed description

I'm trying to view the `sub`s generated for users in a LocalStack Cognito user pool. (The underlying goal is to test JWT-based access control with Hasura [as described here](https://hasura.io/docs/1.0/graphql/core/auth/authentication/jwt.html#the-spec).)

## Expected behavior

When I run `aws cognito-idp list-users` for our production user pool, I get results that look like this (`Username` is the same as the sub attribute):

```
        {
            "Username": "62104cf6-3619-4999-a55a-1ea67b7fa827",
            "Attributes": [
                {
                    "Name": "sub",
                    "Value": "62104cf6-3619-4999-a55a-1ea67b7fa827"
                },
                {
                    "Name": "email",
                    "Value": "x@example.com"
                }
            ],
            "UserCreateDate": "2020-08-07T13:14:40.421000+02:00",
            "UserLastModifiedDate": "2020-08-07T13:18:15.560000+02:00",
            "Enabled": true,
            "UserStatus": "CONFIRMED"
        },
```

Also, the `admin-create-user` response has the `sub` as the username:

```
$ aws cognito-idp admin-create-user --user-attributes "Name=email,Value=foo@example.com" --user-pool-id us-east-2_abc123 --username "foo@example.com"
{
    "User": {
        "Username": "19343459-d76d-41f4-8e04-ca118f2ebdf4",
        "Attributes": [
            {
                "Name": "sub",
                "Value": "19343459-d76d-41f4-8e04-ca118f2ebdf4"
            },
            {
                "Name": "email",
                "Value": "foo@example.com"
            }
        ],
        "UserCreateDate": "2020-09-04T15:06:52.350000-04:00",
        "UserLastModifiedDate": "2020-09-04T15:06:52.350000-04:00",
        "Enabled": true,
        "UserStatus": "FORCE_CHANGE_PASSWORD"
    }
}
```

## Actual behavior

But with LocalStack the results look more like this (`Username` is an email address, and the `sub` attribute is not present at all):

```
        {
            "Username": "x@example.com",
            "Attributes": [
                {
                    "Name": "username",
                    "Value": "x@example.com"
                }
            ],
            "UserCreateDate": 1599553890.0,
            "UserLastModifiedDate": 1599553890.0,
            "Enabled": true,
            "UserStatus": "CONFIRMED",
            "MFAOptions": []
        },
```

And the output of `admin-create-user` has the human username instead of the `sub`:

```
$ awslocal cognito-idp admin-create-user --user-attributes "Name=email,Value=foo@example.com" --user-pool-id us-east-2_498bf40c4c3e44c8a2e4ecb24afdf203 --username "foo@example.com"
{
    "User": {
        "Username": "foo@example.com",
        "Attributes": [
            {
                "Name": "email",
                "Value": "foo@example.com"
            }
        ],
        "UserCreateDate": 1599246553.0,
        "UserLastModifiedDate": 1599246553.0,
        "Enabled": true,
        "UserStatus": "FORCE_CHANGE_PASSWORD",
        "MFAOptions": []
    }
}
```

<details>
<summary>
User pool and client configuration
</summary>

#### User pool

(redacted for brevity)

```
resource "aws_cognito_user_pool" "users" { 
  name = "users"

  auto_verified_attributes = ["email"]
  username_attributes      = ["email"]
  username_configuration { 
    case_sensitive = false
  } 

  lambda_config { 
    pre_token_generation = aws_lambda_function.cognito_pre_token_generation_lambda_trigger.arn
  } 

  admin_create_user_config { 
    allow_admin_create_user_only = true
    } 
  } 

  schema { 
    name                     = "sub"
    attribute_data_type      = "String"
    required                 = true
    mutable                  = false
    developer_only_attribute = false

    string_attribute_constraints { 
      min_length = "1"
      max_length = "2048"
    } 
  } 

  schema { 
    name                = "username"
    attribute_data_type = "String"
  } 

  schema { 
    name                = "email"
    attribute_data_type = "String"
  } 
}
```

#### Client

```
resource "aws_cognito_user_pool_client" "portal-server" {
  name                                 = "portal-server"
  user_pool_id                         = aws_cognito_user_pool.users.id
  generate_secret                      = true
  refresh_token_validity               = 30
  allowed_oauth_flows_user_pool_client = true
  explicit_auth_flows                  = ["ADMIN_NO_SRP_AUTH", "USER_PASSWORD_AUTH", "ALLOW_USER_PASSWORD_AUTH", "ALLOW_USER_SRP_AUTH", "ALLOW_REFRESH_TOKEN_AUTH", "ALLOW_CUSTOM_AUTH", "ALLOW_ADMIN_USER_PASSWORD_AUTH"]
  allowed_oauth_flows                  = ["client_credentials", "code", "implicit"]
  supported_identity_providers         = ["COGNITO"]
  allowed_oauth_scopes                 = ["phone", "email", "openid", "profile", "aws.cognito.signin.user.admin"]
}
```
</details>

## Additional information (sub mismatch)

I noticed that a `sub` is returned from the `awslocal cognito-idp sign-up` command, so at first I thought I could use that:

```
awslocal cognito-idp sign-up --client-id ${client_id} --username foo@example.com --password password
{
    "UserConfirmed": false,
    "UserSub": "27863295-a22c-45a3-bc9e-cc796fd72c5a"
}
```

However, when I retrieve a JWT for this user and decode the `IdToken`, it shows a different sub (`21b0676e-58af-45d1-bb10-77c9796329c0` instead of `27863295-a22c-45a3-bc9e-cc796fd72c5a` as shown above):

<details>
<summary>
Output of `awslocal cognito-idp initiate-auth`
</summary>

```
awslocal cognito-idp initiate-auth --client-id "$client_id" --auth-flow USER_PASSWORD_AUTH --auth-parameters "USERNAME=foo@example.com,PASSWORD=password"
{
    "AuthenticationResult": {
        "AccessToken": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6InRlc3Qta2V5In0.eyJleHAiOjE1OTk3MzAyMDUsImlzcyI6Imh0dHBzOi8vbG9jYWxzdGFjazo0NTkwL3VzLWVhc3QtMl83YWI2ZTg0NzJhNjI0YmE1YTUxYTc5MWVmNjY3NTJlYiIsInN1YiI6IjIxYjA2NzZlLTU4YWYtNDVkMS1iYjEwLTc3Yzk3OTYzMjljMCIsImF1dGhfdGltZSI6MTU5OTY0MzgwNSwiaWF0IjoxNTk5NjQzODA1LCJldmVudF9pZCI6IjA5YmFlOTdhLTMxZTctNDM0Zi1iOWUxLTBjODhiNjQ4ODkyNSIsInRva2VuX3VzZSI6ImFjY2VzcyIsInVzZXJuYW1lIjoiZm9vQGV4YW1wbGUuY29tIiwic2NvcGUiOiJhd3MuY29nbml0by5zaWduaW4udXNlci5hZG1pbiIsImp0aSI6IjA4MTQwNjA1LTQ3Y2UtNGIwNy04MmNlLTNlZWNlMDgwOTczMyIsImNsaWVudF9pZCI6InRucXlza3N5Z2s3ZHMwMXUzYTlieTJ5MmJkIn0.mup1hrirMzEIZ4iQUPZq-o8aXJp9CXhjCLP_9U_w-M2WK87wOVwm3WZ7ixpbz0jS2DEDSsDe7l8xSNX3RW2dm-JmplU3a7keScFXjYwF720Egpba-TmYxcXOdT6SINCyAfqNQw9QdP4AL_UrAHolnDOYJ0LFx_1_oggdIaLoxVP7abXLMUr1gpmdM96iDeMgXJ2Dw7kE-mmL_R3FBVEUexw4cyAbLGaF_bJHAjrSAgOxr0_VN4iVZUCNj_2qTZJHRUXFNjUT3IoXgmvVQlCuM2Wku7nKX3YwlhEbEZCdCOZS6tfWyZm2jMIUVC_YeDoA7VPHoSBT7bVDEmkbfZ0jZg",
        "ExpiresIn": 86400,
        "TokenType": "Bearer",
        "RefreshToken": "944baf56",
        "IdToken": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6InRlc3Qta2V5In0.eyJleHAiOjE1OTk3MzAyMDUsImlzcyI6Imh0dHBzOi8vbG9jYWxzdGFjazo0NTkwL3VzLWVhc3QtMl83YWI2ZTg0NzJhNjI0YmE1YTUxYTc5MWVmNjY3NTJlYiIsInN1YiI6IjIxYjA2NzZlLTU4YWYtNDVkMS1iYjEwLTc3Yzk3OTYzMjljMCIsImF1dGhfdGltZSI6MTU5OTY0MzgwNSwiaWF0IjoxNTk5NjQzODA1LCJldmVudF9pZCI6ImE0MjFmYzgxLTk1MTUtNDkxMy1iZGYxLWFkMDA1ZjRjZWJjMCIsInRva2VuX3VzZSI6ImlkIiwiYXVkIjoidG5xeXNrc3lnazdkczAxdTNhOWJ5MnkyYmQiLCJjb2duaXRvOnVzZXJuYW1lIjoiZm9vQGV4YW1wbGUuY29tIiwiaHR0cHM6Ly9oYXN1cmEuaW8vand0L2NsYWltcyI6IntcIlgtSGFzdXJhLUFsbG93ZWQtUm9sZXNcIjpbXCJ1c2VyXCJdLFwiWC1IYXN1cmEtRGVmYXVsdC1Sb2xlXCI6XCJ1c2VyXCIsXCJYLUhhc3VyYS1Vc2VyLUlkXCI6XCIyMWIwNjc2ZS01OGFmLTQ1ZDEtYmIxMC03N2M5Nzk2MzI5YzBcIn0ifQ.mroO1f82N0ZZ_9bMO_OvqtRLQy1qHr8SubGTBuLWDlmK4uq28XUa_BkSEsfYSoDaXJ9TRkrxUTWoQmtCQjuGiVlhi_lFlyrskPn946koflcCFt4osbcUP_el9VIbJqXcOdvyjl2ba0Sa4iZmqFNgbzRO_qht3YyK-MVZrpPzRjH0s94As1SfT43nMg3Q8nlvvlot0V5njnTceV4FspUeNvgl1tlq4z5XvlKec_QTHhc67m-ELOc6ihyIcBk4EduiXqN_4tEhVnHgSiGtbTMP6DsoZT0WkQhvoKMq4sIOLCTxw561qGB9dXjluRlqmvFb9P1Nx4u__AASHj4VGJLzNQ"
    }
}
```
</details>

Decoded:

```
{
  "exp": 1599730205,
  "iss": "https://localstack:4590/us-east-2_7ab6e8472a624ba5a51a791ef66752eb",
  "sub": "21b0676e-58af-45d1-bb10-77c9796329c0",
  "auth_time": 1599643805,
  "iat": 1599643805,
  "event_id": "a421fc81-9515-4913-bdf1-ad005f4cebc0",
  "token_use": "id",
  "aud": "tnqysksygk7ds01u3a9by2y2bd",
  "cognito:username": "foo@example.com",
  "https://hasura.io/jwt/claims": "{\"X-Hasura-Allowed-Roles\":[\"user\"],\"X-Hasura-Default-Role\":\"user\",\"X-Hasura-User-Id\":\"21b0676e-58af-45d1-bb10-77c9796329c0\"}"
}
```

TL;DR How can one obtain the user `sub`s?

**Comment by whummer** [comment]
Thanks for the detailed report @soulshake . As discussed via Slack, this should (hopefully) be fixed in the `latest` version. 

We're following the logic defined in the [Cognito docs](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html). Basically, if the user pool is created with `UsernameAttributes=[...]` (specifying 'email' and/or 'phone_number'), then the `Username` should automatically be set to the `sub` value, and the username itself would get added as an `email` or `phone_number` attribute on the Cognito user itself.

Can you please give it a try and let us know if the problem persists. Thanks

Also - Does the recent change in `latest` also fix the issue reported in #2964 ? Thanks for your help.

**Comment by soulshake** [comment]
@whummer Unfortunately it doesn't look like this resolves https://github.com/localstack/localstack/issues/2964. I updated that issue with the latest output (see "Example event as received by Lambda running on LocalStack").


I am not sure if they were directly caused by this change but there are a couple of new issues related to this one:

- https://github.com/localstack/localstack/issues/2995
- https://github.com/localstack/localstack/issues/2996
